zebra-state/src/request.rs

```diff
@@ -53,6 +54,22 @@ impl From<block::Height> for HashOrHeight {
     }
 }
 
+impl std::str::FromStr for HashOrHeight {
+    type Err = SerializationError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.parse() {
+            Ok(hash) => Ok(Self::Hash(hash)),
+            Err(_) => match s.parse() {
+                Ok(height) => Ok(Self::Height(height)),
+                Err(_) => Err(SerializationError::Parse(
+                    "Could not convert the input string to a hash nor a height",
+                )),
```

Comment:
This is very neat code, I really like the type inference here!

Reply:
One alternative is [`Result::or_else`](https://doc.rust-lang.org/std/result/enum.Result.html#method.or_else), something like:

```rust
s.parse()
    .map(Self::Hash)
    .or_else(|_error| s.parse().map(Self::Height))
    .map_err(|error| SerializationError::Parse("Could not convert the input string to a hash nor a height"))
```

Reply:
If it wants a `'static &str`, I don't think it's possible due to lifetimes.

Reply:
Yes, the definition is `Parse(&'static str)`.

I'll leave it as it is then.